src/decoder/mod.rs

```diff
@@ -28,7 +27,25 @@ pub enum DecodingResult {
 }
 
 impl DecodingResult {
-    pub fn to_buffer(&mut self, start: usize) -> DecodingBuffer {
+    fn new_u8(size: usize, limits: &Limits) -> TiffResult<DecodingResult> {
+        if size as u64 > limits.decoding_buffer_size {
+            Err(TiffError::LimitsExceeded)
+        }
+        else {
+            Ok(DecodingResult::U8(vec![0; size]))
+        }
+    }
+
+    fn new_u16(size: usize, limits: &Limits) -> TiffResult<DecodingResult> {
+        if size as u64*2 > limits.decoding_buffer_size {
```

Comment:
This one should probably be `size as u64 > limits.decoding_buffer_size/2` otherwise the mult may overflow, or use `checked_mul` instead.

Reply:
Overall, I'm not sure if the case of `usize == u32` is handled properly for all the multiplications everywhere (i.e. doesn't overflow them).